phantom-streams/src/main/scala/com/outworkers/phantom/streams/lib/EnumeratorSubscription.scala

```diff
@@ -0,0 +1,218 @@
+  /**
+    * The Iteratee state before any elements have been requested, before
+    * we've attached an Iteratee to the Enumerator.
+    */
+  final case object Unattached extends IterateeState[Nothing]
+  /**
+    * The Iteratee state when we're reading from the Enumerator.
+    */
+  final case class Attached[T](link: Promise[Iteratee[T, Unit]]) extends IterateeState[T]
+
+}
+
+
+/**
+  * Adapts an Enumerator to a Publisher.
+  */
+private[streams] class EnumeratorSubscription[T, U >: T](
+  enum: Enumerator[T],
+  emptyElement: Option[T],
+  subr: Subscriber[U],
+  onSubscriptionEnded: SubscriptionHandle[U] => Unit)
+  extends StateMachine[State[T]](initialState = Requested[T](0, Unattached))
+    with Subscription with SubscriptionHandle[U] {
+
+  // SubscriptionHandle methods
+
+  override def start(): Unit = {
+    subr.onSubscribe(this)
+  }
+
+  override def subscriber: Subscriber[U] = subr
+
+  override def isActive: Boolean = {
+    // run immediately, don't need to wait for exclusive access
+    state match {
+      case Requested(_, _) => true
+      case Completed | Cancelled => false
+    }
+  }
+
+  // Streams methods
+  override def request(elements: Long): Unit = {
+    if (elements <= 0) throw new IllegalArgumentException(s"The number of requested elements must be > 0: requested $elements elements")
+    exclusive {
+      case Requested(0, its) =>
+        state = Requested(elements, extendIteratee(its))
+      case Requested(n, its) =>
+        state = Requested(n + elements, its)
+      case Completed | Cancelled =>
+        ()
+    }
+  }
+
+  override def cancel(): Unit = exclusive {
+    case Requested(_, its) =>
+      val cancelLink: Iteratee[T, Unit] = Done(())
+      its match {
+        case Unattached =>
+          enum(cancelLink)
+        case Attached(link0) =>
+          link0.success(cancelLink)
+      }
+      state = Cancelled
+    case Cancelled | Completed =>
+      ()
+  }
+
+  // Methods called by the iteratee when it receives input
+
+  /**
+    * Called when the Iteratee received Input.El, or when it recived
+    * Input.Empty and the Publisher's `emptyElement` is Some(el).
+    */
+  private def elementEnumerated(el: T): Unit = exclusive {
+    case Requested(1, its) =>
+      subr.onNext(el)
+      state = Requested(0, its)
+    case Requested(n, its) =>
+      subr.onNext(el)
+      state = Requested(n - 1, extendIteratee(its))
+    case Cancelled =>
+      ()
+    case Completed =>
+      throw new IllegalStateException("Shouldn't receive another element once completed")
+  }
+
+  /**
+    * Called when the Iteratee received Input.Empty and the Publisher's
+    * `emptyElement` value is `None`
+    */
+  private def emptyEnumerated(): Unit = exclusive {
+    case Requested(n, its) =>
+      state = Requested(n, extendIteratee(its))
+    case Cancelled =>
+      ()
+    case Completed =>
+      throw new IllegalStateException("Shouldn't receive an empty input once completed")
```

Comment:
Have you every considered using this pattern?
```
case object EmptyInputOnceCompleted extends IllegalStateException("Shouldn't receive an empty input once completed")
```
It allows refined case matching of the exception on the invocation side :)